### ISSUE TITLE
On-device initialisaiton functions (or step/exit functions?)

### ISSUE THREAD

**Issue by ptheywood** [FLAMEGPU/FLAMEGPU2_dev#329]
Current init functions are sequential host methods. 

In some cases, doing this on the device is probably worthwhile (although it adds complexity). 

General idea is:

+ `intiLayers` as a set of layers which contain host or device initialisation functions (or submodels)
+ the majority of the body of `step` becomes a private method which takes an ordered list of layers and processes them
    + `step` then calls this method for init, agent, step and exit layer sets etc.
+ The method to add a host init (/step/exit) function, either adds to an exisitng layer, or implicitly creates and appends a new layer (if a layer is not provided)
+ Restrictions will have to be imposed on which functions can be in the same layer(s) for the sake of concurrency, as with restrictiosn on layers.

Longer term the layers would become a DAG as with the main layers.

Would need to make sure curand behaves as intended, and potentially provide simple methods for provided non-uniform distributions across the populations (i.e normal distributions) otherwise a sequential host method would still have to be used instead.

**Comment by Robadob** [comment]
Simulate then calls this method...
And not for step functions. That would create unwanted recursion

**Comment by Robadob** [comment]
> a public version of `step` is still required though which doesn't take a set of layers though surely? I.e. where you call step explicilty in the test suite?

Yes, `step()` currently never calls init or exit fns, hence it should probably read: "`simulate` then calls this method for init, agent, step and exit layer sets etc."


> I don't see how it would be recursion, Step functions are just methods executed after the agent function layers, so no reason why they can't also have their own set of layers (though arguably there shoulnt be agent functions in there, but certainly scope for device step functions in some very obscure cases)?

Yeah, I was mistaken on this point whilst reading and walking. I was thinking a vector of layers, representing a step would have step functions.

> Argubaly we don't actually need explicit step functions anymore, as we have host-layer functions, so step functions are just host layer functions which are implicitly at the end of the iteration.

I agree, but they exist as backwards compatibility.

**Comment by Robadob** [comment]
This alludes to device-agent initialisation functions (and they don't have a separate issue).

The easiest way to handle this would be with the existing Device API and agent function wrapper used by regular agent functions. But that would leave unusable message in/out properties (they would be treated as MsgNone) and redundant? read access to agent variables. The alternative being to make a bespoke InitDeviceApi, which then leads to either duplicated code or double inheritance/similar.

**Comment by Robadob** [comment]
Init/step/exit layers seems like a priority for Primage initialisation performance, so this may get addressed soon.
